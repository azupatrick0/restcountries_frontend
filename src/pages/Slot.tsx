/* eslint-disable react-hooks/exhaustive-deps */
import React, { FC, useState } from 'react';
import { useSelector, useDispatch, shallowEqual } from 'react-redux';
import { uuid } from 'uuidv4';
import NavBar from 'components/NavBar';
import SpinCard from 'components/Card/SpinCard';
import Loader from 'components/Loader';
import spin from 'actions/spin';
import Button from 'components/Button';

export interface iState {
	spin: any;
	loading: boolean;
	error: string;
}

export const Slot: FC = () => {
  const [balance, setBalance] = useState('');
	const state = useSelector((state: iState) => state, shallowEqual);
  const dispatch = useDispatch();
  
  React.useEffect(() => {
    if (localStorage.getItem('balance') === null) {
      localStorage.setItem('balance', '20');
    } else if (state.spin.spin.won) {
      const oldBalance:any = localStorage.getItem('balance');
      const newBalance = state.spin.spin.won > 0 ? `${(+oldBalance + 1) + state.spin.spin.won}` : `${+oldBalance + state.spin.spin.won}`;
      localStorage.setItem('balance', newBalance);
      setBalance(newBalance);
    }
  }, [state.spin.spin.won]);

	const setCustomerBalance = () => {
    let balance:any = localStorage.getItem('balance') && localStorage.getItem('balance');
    balance = +balance - 1;
    localStorage.setItem('balance', `${balance}`);
    setBalance(`${balance}`);
	};

	const handleSpin = () => {
    dispatch(spin());
    setCustomerBalance();
	};

	return (
		<div className="spin">
			<NavBar
        balance={balance || 20}
        won={state.spin.spin.won ? state.spin.spin.won : 0}
			/>
      <div className="spin__balance">
        {
          balance === '0' ? 
          <div className="spin__balance-view">
            <span></span>
            <span className="balance__depleted">Balance depleted, please fund again</span>
            <span></span>
          </div>
          :
          <div className="spin__button-view">
            <span></span>
            <Button
              onClick={handleSpin}
              className={"spin__button-view-button"}
            >
              SPIN
            </Button>
            <span></span>
          </div>
          
        }
      </div>
			<div className="cards__layout">
				{state.spin.loading && state.spin.loading ? (
					<Loader />
				) : state.spin.spin.result && state.spin.spin.result.length ? (
					state.spin.spin.result.map((spinResult: string): any => (
						<SpinCard
              spin={spinResult}
              id={uuid()}
              key={uuid()}
						/>
					))
				) : (
					'No spin yet, please click the spin button'
				)}
			</div>
		</div>
	);
};

export default Slot;
