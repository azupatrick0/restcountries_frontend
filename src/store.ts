import { createStore, applyMiddleware, combineReducers } from 'redux';
import { composeWithDevTools } from 'redux-devtools-extension';
import thunk from 'redux-thunk';
import fetchAllCountries from 'reducers/fetchAllCountries';

const rootReducer = combineReducers({
  countries: fetchAllCountries,
});

const store = createStore(rootReducer, composeWithDevTools(applyMiddleware(thunk)));

export default store;