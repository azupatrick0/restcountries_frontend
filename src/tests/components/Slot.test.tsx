import React from 'react';
import { mount } from 'enzyme';
import Slot from 'pages/Slot';
import { MemoryRouter } from 'react-router-dom';
import { Provider } from 'react-redux';
import store from 'store';

describe("Slot test suite", () => {
	it('should render slot page', () => {
		const wrapper = mount(
			<MemoryRouter>
				<Provider store={store}>
					<Slot />
				</Provider>
			</MemoryRouter>
		);

		expect(wrapper.find('.balance').find('div').at(0).text()).toContain('Balance');
		expect(wrapper).toMatchSnapshot();
	});

	it('should return spin result when spin button is clicked', async () => {
		const wrapper = mount(
			<MemoryRouter>
				<Provider store={store}>
					<Slot />
				</Provider>
			</MemoryRouter>
		);

    wrapper.find('.spin__button-view-button').at(1).simulate('click');

    expect(wrapper.find('.cards__layout').find('div').length).toBeGreaterThan(0);
    expect(Number(wrapper.find('.balance').find('.won').find('span').text())).toBeGreaterThanOrEqual(0);
	});
});
