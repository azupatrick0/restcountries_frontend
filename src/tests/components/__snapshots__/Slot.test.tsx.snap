// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Slot test suite should render slot page 1`] = `ReactWrapper {}`;
